### PR TITLE
#933 Cells should be disabled when a document is -- You won't believe how many files were touched for this ONE SIMPLE CHANGE -- [14]

### LINKED ISSUE #933: The disabled state of a row should be passed onto the cells
Currently we don't disable editable cells for anything.

We should! - finalised etc.

The `DataRow` component has a disabled state which can be set via the table context.

We should pass the disabled state to the cells, by adding `isDisabled: boolean` to the `Cell<T>` type, and then within each cell component, handle the disabled state

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #933 

**I've realised that maybe a lot of this is all wrong! (Disabling of inputs in edit modals). Maybe it would just be better to disable opening the modal rather than disabling it..! That would certainly be a lot easier..**

- Added a field to `CellProps<T>`, `isDisabled`. Which `DataRow` is able to pass into cells.
- Updated each of the cells/columns to have a disabled state.
- Then, also went and triggered a disabled state in the list views and detail modals - but NOT outbound, will make a separate issue for that.
- Also need a separate issue for stocktake edit modal in a finalised state. Should show only the existing stocklines, rather than all of the stock lines available to count.

### CODE REVIEW

File: packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx [hunk 1]
```diff
@@ -8,12 +8,15 @@ export const BaseDatePickerInput: FC<
...
       renderInput={(params: TextFieldProps) => {
-        const props: StandardTextFieldProps = {
+        const textInputProps: StandardTextFieldProps = {
           ...params,
           variant: 'standard',
         };
-        return <BasicTextInput {...props} />;
+        return (
+          <BasicTextInput disabled={!!props.disabled} {...textInputProps} />
+        );
```
Comment (joshxg) [inline packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:19]: Adding disabled

File: packages/common/src/ui/components/inputs/DatePickers/ExpiryDateInput/ExpiryDateInput.tsx [hunk 2]
```diff
@@ -5,14 +5,17 @@ import lastDayOfMonth from 'date-fns/lastDayOfMonth';
...
 }
 
 export const ExpiryDateInput: FC<ExpiryDateInputProps> = ({
   value,
   onChange,
+  disabled,
 }) => {
   return (
     <BaseDatePickerInput
+      disabled={disabled}
```
Comment (joshxg) [inline packages/common/src/ui/components/inputs/DatePickers/ExpiryDateInput/ExpiryDateInput.tsx:18]: Adding disabled

File: packages/common/src/ui/layout/tables/DataTable.tsx [hunk 3]
```diff
@@ -29,13 +29,25 @@ export const DataTableComponent = <T extends RecordWithId>({
...
   }, [data]);
 
+  useEffect(() => {
+    console.log(
+      'setting disabled',
+      data.map(({ id }) => id)
+    );
+    if (isDisabled) setDisabledRows(data.map(({ id }) => id));
+  }, [isDisabled, data]);
+
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/DataTable.tsx:50]: Not sure if this is the best idea, but.. it was easy to implement!

Essentially passing `isDisabled` into the `DataTable` disables all rows

File: packages/inventory/src/Stocktake/DetailView/modal/StocktakeLineEdit/StocktakeLineEdit.tsx [hunk 4]
```diff
@@ -8,6 +8,8 @@ import {
   Box,
   ModalMode,
   useNotification,
+  TableProvider,
```
Comment (joshxg) [inline packages/inventory/src/Stocktake/DetailView/modal/StocktakeLineEdit/StocktakeLineEdit.tsx:11]: This is updating the StocktakeLineEdit modal to be disabled when finalised. 

Adding a `TableProvider` around the modal / tables and then just passing `isDisabled` into the `DataTable`

File: packages/inventory/src/Stocktake/ListView/ListView.tsx [hunk 5]
```diff
@@ -7,13 +7,22 @@ import {
...
 import { StocktakeRowFragment } from '../api/operations.generated';
 import { useStocktakes } from '../api';
 
+const useDisableStocktakeRows = (rows?: StocktakeRowFragment[]) => {
+  const { setDisabledRows } = useTableStore();
+  useEffect(() => {
+    const disabledRows = rows?.filter(isStocktakeDisabled).map(({ id }) => id);
+    if (disabledRows) setDisabledRows(disabledRows);
+  }, [rows]);
+};
```
Comment (joshxg) [inline packages/inventory/src/Stocktake/ListView/ListView.tsx:24]: disabling 'disabled' rows - i.e. locked and/or finalised.

Has the potential disadvantage that the row is grey/gray/grae - though it's a bit of feedback I suppose

File: packages/invoices/src/InboundShipment/DetailView/AppBarButtons.tsx [hunk 6]
```diff
@@ -7,7 +7,7 @@ import {
   useDetailPanel,
   useTranslation,
 } from '@openmsupply-client/common';
-import { useIsInboundEditable } from '../api';
+import { useIsInboundDisabled } from '../api';
```
Comment (joshxg) [inline packages/invoices/src/InboundShipment/DetailView/AppBarButtons.tsx:10]: From here down is the same as stocktake.

Add a hook to list views to disable 'disabled' rows.

File: packages/invoices/src/InboundShipment/DetailView/AppBarButtons.tsx [hunk 7]
```diff
@@ -16,15 +16,15 @@ interface AppBarButtonProps {
 export const AppBarButtonsComponent: FC<AppBarButtonProps> = ({
   onAddItem,
 }) => {
-  const isEditable = useIsInboundEditable();
+  const isDisabled = useIsInboundDisabled();
```
Comment (joshxg) [inline packages/invoices/src/InboundShipment/DetailView/AppBarButtons.tsx:19]: renamed `Editable` -> `isDisabled` finally

File: packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEdit.tsx [hunk 8]
```diff
@@ -167,140 +171,146 @@ export const InboundLineEdit: FC<InboundLineEditProps> = ({
...
-    >
-      {isLoading ? (
-        <BasicSpinner messageKey="saving" />
-      ) : (
-        <>
-          <InboundLineEditForm
-            disabled={mode === ModalMode.Update}
-            item={currentItem}
-            onChangeItem={setCurrentItem}
+    <TableProvider createStore={createTableStore}>
```
Comment (joshxg) [inline packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEdit.tsx:174]: for disabling inbound rows

File: packages/requisitions/src/RequestRequisition/ListView/ListView.tsx [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import React, { FC, useCallback } from 'react';
+import React, { FC, useCallback, useEffect } from 'react';
```
Comment (joshxg) [inline packages/requisitions/src/RequestRequisition/ListView/ListView.tsx:1]: requisitions don't have a colour (I reckon they should.. why not?!) so the row is disabled for deleting, but the gray seems even more wrong here

**Comment by mark-prins** [comment]
Added a couple in d97d7fd8
The stocktake modal was blowing up due to the `CheckboxCell` and the stocktake was unable to be locked

**Review by mark-prins — APPROVED** [review]
🎉  and a new record for the number of files changed for a simple requirement!!

worth it though - better to have the modal open and be disabled than to block opening the modal.
Popped in a sneaky commit to fix some issues, they might be ok in other PRs.. if so, it was only a couple of lines, so should be ok to undo.

**Pull request merged**